network/health_check.go

```diff
@@ -0,0 +1,155 @@
+	logger := logging.FromContext(ctx)
+	port := os.Getenv("KNATIVE_HEALTH_PROBES_PORT")
+	if port == "" {
+		port = "8080"
+	}
+
+	server := http.Server{ReadHeaderTimeout: time.Minute, Handler: muxWithHandles(ctx), Addr: ":" + port}
+
+	go func() {
+		go func() {
+			<-ctx.Done()
+			_ = server.Shutdown(ctx)
+		}()
+
+		// start the web server on port and accept requests
+		logger.Infof("Probes server listening on port %s", port)
+
+		if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+			logger.Fatal(err)
+		}
+	}()
+}
+
+func muxWithHandles(ctx context.Context) *http.ServeMux {
+	var handler HealthCheckHandler
+	if h := getOverrideHealthCheckHandler(ctx); h != nil {
+		handler = *h
+	} else {
+		defaultHandle := newDefaultProbesHandle(ctx)
+		handler = HealthCheckHandler{
+			ReadinessProbeRequestHandle: defaultHandle,
+			LivenessProbeRequestHandle:  defaultHandle,
+		}
+	}
+	mux := http.NewServeMux()
+	if UserReadinessProbe(ctx) {
+		mux.HandleFunc("/readiness", handler.ReadinessProbeRequestHandle)
+	}
+	if UserLivenessProbe(ctx) {
+		mux.HandleFunc("/health", handler.LivenessProbeRequestHandle)
+	}
+	// Set both probes if user does not want to explicitly set only one.
+	if !UserReadinessProbe(ctx) && !UserLivenessProbe(ctx) {
+		mux.HandleFunc("/readiness", handler.ReadinessProbeRequestHandle)
+		mux.HandleFunc("/health", handler.LivenessProbeRequestHandle)
+	}
+	return mux
+}
+
+// HealthCheckHandler allows to set up a handler for probes. User can override the default handler
+// with a custom one.
+type HealthCheckHandler struct {
+	ReadinessProbeRequestHandle http.HandlerFunc
+	LivenessProbeRequestHandle  http.HandlerFunc
+}
+
+func newDefaultProbesHandle(sigCtx context.Context) http.HandlerFunc {
+	logger := logging.FromContext(sigCtx)
+	once := sync.Once{}
+	return func(w http.ResponseWriter, r *http.Request) {
+		f := func() error {
+			select {
+			// When we get SIGTERM (sigCtx done), let readiness probes start failing.
+			case <-sigCtx.Done():
+				once.Do(func() {
+					logger.Info("Signal context canceled")
+				})
+				return errors.New("received SIGTERM from kubelet")
+			default:
+				return nil
+			}
+		}
+
+		if IsKubeletProbe(r) {
+			if err := f(); err != nil {
+				logger.Errorf("Healthcheck failed: %v", err)
+				http.Error(w, err.Error(), http.StatusInternalServerError)
+			} else {
+				w.WriteHeader(http.StatusOK)
+			}
+			return
+		}
+		http.Error(w, "Unexpected request", http.StatusBadRequest)
+	}
+}
+
+type addUserReadinessProbeKey struct{}
+
+// WithUserReadinessProbe signals to ServeHealthProbes that it should set explicitly a readiness probe.
+func WithUserReadinessProbe(ctx context.Context) context.Context {
+	return context.WithValue(ctx, addUserReadinessProbeKey{}, struct{}{})
+}
+
+// UserReadinessProbe checks if user has explicitly requested to set a readiness probe in the related context.
+func UserReadinessProbe(ctx context.Context) bool {
```

Comment:
We can extend with a custom url in the future to change the default path too.

Reply:
I'm struggling to figure out why you'd want to disable these on the server.  I understand why you might want to remove them from the manifest, but what's the rationale for not registering them with the HTTPServer?